### PR TITLE
[gpu] Unused parameter 

### LINKED ISSUE #5036: [gpu] Unused parameter
I don't know why the line numbers are wrong, but I'm getting following warnings:

```cpp
/.../pcl/gpu/containers/src/initialization.cpp:100:86: warning: unused parameter 'device_attribute' [-Wunused-parameter]
    template <class T> inline void getCudaAttribute(T *attribute, CUdevice_attribute device_attribute, int device)
                                                                                     ^
/.../pclgpu/containers/src/initialization.cpp:100:108: warning: unused parameter 'device' [-Wunused-parameter]
    template <class T> inline void getCudaAttribute(T *attribute, CUdevice_attribute device_attribute, int device)
                                                                                                           ^
/.../pcl/gpu/containers/src/error.cpp:42:94: warning: unused parameter 'func' [-Wunused-parameter]
void pcl::gpu::error(const char *error_string, const char *file, const int line, const char *func)
                                                                                             ^
```

### PR DESCRIPTION
**Pull request by xiaolong-666**
Closes #5036 
In error.cpp, deleted unused parameter `func`
In initialization.cpp, enable `cuDeviceGetAttribute` again.

### CODE REVIEW

**Comment by larshg** [comment]
It wants all arguments for `gpu::error` on one line see:
https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=20488&view=logs&j=84ba2080-ade3-55e3-f209-94ce73f9d406&t=9838d42f-a715-5615-242f-7e73bf03078b

File: gpu/containers/src/error.cpp [hunk 1]
```diff
@@ -42,8 +42,7 @@
 void
 pcl::gpu::error(const char* error_string,
                 const char* file,
-                const int line,
-                const char* func)
+                const int line)
```
Comment (kunaltyagi) [inline gpu/containers/src/error.cpp:45]:
```suggestion
pcl::gpu::error(const char* error_string, const char* file, const int line)
```

**Review by kunaltyagi — COMMENTED** [review]
LGTM. Looking into the reasons requested by mvieth

**Comment by larshg** [comment]
You need to update here as well:

https://github.com/PointCloudLibrary/pcl/blob/9e576210221fbee27599542d55bff798109d984a/gpu/containers/include/pcl/gpu/containers/initialization.h#L77-L83

**Comment by xiaolong-666** [comment]
> You need to update here as well:
> 
> https://github.com/PointCloudLibrary/pcl/blob/9e576210221fbee27599542d55bff798109d984a/gpu/containers/include/pcl/gpu/containers/initialization.h#L77-L83

Why should I delete it here? Doesn't it provide default values?

**Comment by larshg** [comment]
Your method should have the same number of parameters in both .cpp and .h file. However, in the .h file you can supply default arguments yes.

**Comment by mvieth** [comment]
@xiaolong-666 May I ask why you deleted the parameter `func` instead of using it in the `std::cout` print statement? That could improve the error message, and changing the parameters of an existing function is always difficult for compatibility reasons

**Comment by xiaolong-666** [comment]
> @xiaolong-666 May I ask why you deleted the parameter `func` instead of using it in the `std::cout` print statement? That could improve the error message, and changing the parameters of an existing function is always difficult for compatibility reasons


I checked wherever it was called and only the first three parameters were used, so I deleted it.

**Comment by mvieth** [comment]
> @mvieth I checked wherever it was called and only the first three parameters were used, so I deleted it.

I found three places where also the `func` parameter is used:
https://github.com/PointCloudLibrary/pcl/blob/master/gpu/kinfu/src/safe_call.hpp#L56
https://github.com/PointCloudLibrary/pcl/blob/master/gpu/utils/include/pcl/gpu/utils/safe_call.hpp#L56
https://github.com/PointCloudLibrary/pcl/blob/master/gpu/people/src/cuda/utils.cu#L94
(Plus it is in the public API, so there is always a small possibility that someone outside of the library uses it, even if the doc says "for internal use only")

**Comment by xiaolong-666** [comment]
> > @mvieth我检查了它被调用的地方，只使用了前三个参数，所以我删除了它。
> 
> 我发现了三个`func`使用参数的地方： [https](https://github.com/PointCloudLibrary/pcl/blob/master/gpu/kinfu/src/safe_call.hpp#L56) : [//github.com/PointCloudLibrary/pcl/blob/master/gpu/kinfu/src/safe_call.hpp#L56 ](https://github.com/PointCloudLibrary/pcl/blob/master/gpu/kinfu/src/safe_call.hpp#L56) [https://github.com/PointCloudLibrary/pcl/ blob/master/gpu/utils/include/pcl/gpu/utils/safe_call.hpp#L56 ](https://github.com/PointCloudLibrary/pcl/blob/master/gpu/utils/include/pcl/gpu/utils/safe_call.hpp#L56) https://github.com/PointCloudLibrary/pcl/blob/master/gpu/people/src/cuda/utils.cu#L94 （另外它在公共 API 中，所以图书馆外的人使用它的可能性很小，即使文档说“仅供内部使用”）

@mvieth I checked it with the keyword `pcl:gpu:error`, I will change it back soon, sorry

**Comment by mvieth** [comment]
No problem!
Unfortunately, the formatting check failed again: https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=20500&view=logs&j=84ba2080-ade3-55e3-f209-94ce73f9d406&t=9838d42f-a715-5615-242f-7e73bf03078b

**Comment by xiaolong-666** [comment]
> No problem! Unfortunately, the formatting check failed again: https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=20500&view=logs&j=84ba2080-ade3-55e3-f209-94ce73f9d406&t=9838d42f-a715-5615-242f-7e73bf03078b

@mvieth  `cuDeviceGetAttribute' doesn't seem to be enabled,  caused the build to fail: https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=20507&view=logs&j=149f4e41-9368-548e-d6d4-ff21d16f0ace&t=30234468-9402-53e3-1a03-9eb9da598fc6
so, what should I do?

**Comment by larshg** [comment]
I have looked a bit into this. And we need to link towards -lcuda. On my machine its in the `CUDA_CUDA_LIBRARY` variable. For some reason, when just enabling CUDA with `enable_language(cuda)` it is not linking to this "Driver" library as it seems called.

So in the CMakelists.txt of the gpu/container project you need to add a link_libraries(NameOfProject ${CUDA_CUDA_LIBRARY}) or something similar. I haven't checked if it is found in the same variable on linux, but hopefully.

**Comment by xiaolong-666** [comment]
> I have looked a bit into this. And we need to link towards -lcuda. On my machine its in the `CUDA_CUDA_LIBRARY` variable. For some reason, when just enabling CUDA with `enable_language(cuda)` it is not linking to this "Driver" library as it seems called.
> 
> So in the CMakelists.txt of the gpu/container project you need to add a target_link_libraries(NameOfProject ${CUDA_CUDA_LIBRARY}) or something similar. I haven't checked if it is found in the same variable on linux, but hopefully.

Thanks, but it might be a little difficult for me, can you help me? I‘m a beginner.

**Comment by mvieth** [comment]
Have you tried what Lars suggested:

> So in the CMakelists.txt of the gpu/container project you need to add a target_link_libraries(NameOfProject ${CUDA_CUDA_LIBRARY})

NameOfProject would be `${LIB_NAME}` here

**Comment by xiaolong-666** [comment]
> Have you tried what Lars suggested:
> 
> > So in the CMakelists.txt of the gpu/container project you need to add a target_link_libraries(NameOfProject ${CUDA_CUDA_LIBRARY})
> 
> NameOfProject would be `${LIB_NAME}` here

emm,  `${CUDA_CUDA_LIBRARY}`  not found in the same variable on linux

**Comment by mvieth** [comment]
Can you try `CUDA_LIBRARIES`?
Alternatively, `set_source_files_properties` on initialization.cpp might work?

**Comment by xiaolong-666** [comment]
> Can you try `CUDA_LIBRARIES`? Alternatively, `set_source_files_properties` on initialization.cpp might work?

Hi, @mvieth The first method failed, so I tried the second. But 
I don't know `PROPERTIES` What to write after, like `set_source_files_properties(src/initialization.cpp PROPERTIES ??)`

**Comment by mvieth** [comment]
> But I don't know `PROPERTIES` What to write after, like `set_source_files_properties(src/initialization.cpp PROPERTIES ??)`

Check out the other place where `set_source_file_properties` is used in gpu/containers/CMakeLists.txt

**Comment by mvieth** [comment]
@larshg Do you have another idea how to get this to compile?
If not, I would suggest to keep `cuDeviceGetAttribute` disabled, and instead comment the unused parameters (like `/*parameter*/`) to silence the warnings

**Comment by mvieth** [comment]
@xiaolong-666 Since we can't find a way to make it compile, I think we should go with my other suggestion from earlier:

> I would suggest to keep `cuDeviceGetAttribute` disabled, and instead comment the unused parameters (like `/*parameter*/`) to silence the warnings

Would you make the changes?

**Comment by xiaolong-666** [comment]
> /*parameter*/



> @xiaolong-666 Since we can't find a way to make it compile, I think we should go with my other suggestion from earlier:
> 
> > I would suggest to keep `cuDeviceGetAttribute` disabled, and instead comment the unused parameters (like `/*parameter*/`) to silence the warnings
> 
> Would you make the changes?



> @xiaolong-666 Since we can't find a way to make it compile, I think we should go with my other suggestion from earlier:
> 
> > I would suggest to keep `cuDeviceGetAttribute` disabled, and instead comment the unused parameters (like `/*parameter*/`) to silence the warnings
> 
> Would you make the changes?

I tried it

**Comment by xiaolong-666** [comment]
> @xiaolong-666 Since we can't find a way to make it compile, I think we should go with my other suggestion from earlier:
> 
> > I would suggest to keep `cuDeviceGetAttribute` disabled, and instead comment the unused parameters (like `/*parameter*/`) to silence the warnings
> 
> Would you make the changes?

@mvieth I don't know why the format check is not passing without any hint

**Comment by mvieth** [comment]
You can see the formatting check [here](https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=21066&view=logs&jobId=84ba2080-ade3-55e3-f209-94ce73f9d406&j=84ba2080-ade3-55e3-f209-94ce73f9d406&t=9838d42f-a715-5615-242f-7e73bf03078b). You can also navigate there yourself, by clicking on "Details" next to the failing check, then click on "View more details on Azure Pipelines", and finally on "Show Diff".
Also: the current state of the code does not fix the unused parameter warning. I had something like this in mind: `getCudaAttribute(T* attribute, CUdevice_attribute /*device_attribute*/, int /*device*/)`

**Review by mvieth — APPROVED** [review]
Thanks!

**Pull request merged**